### ISSUE TITLE
ClientSettingsEvent doesn't work for custom locales.

### ISSUE THREAD

**Issue by minecreatr** [SpongePowered/SpongeAPI#1510]
If a client connects to a server with a custom localization, or changes their localization to one, the ClientSettingsEvent will not be called, as well as an error will be printed to the console. The main fix I can see would be to not force languages to conform to a preset list, but I wanted to report this as it was preventing me from implementing a major feature in a plugin that involves custom languages.

**Comment by kashike** [comment]
>as well as an error will be printed to the console

You know, it would help if you were to provide said error.

> not force languages to conform to a preset list

We don't. We simply expect locales to match the expected pattern.

**Comment by minecreatr** [comment]
Im sorry about forgetting to provide the error, here it is though http://pastebin.com/FxUecj7Y .

**Comment by kashike** [comment]
Yeah, we aren't going to support random strings.
- The language code must be lowercase.
- The country code must be uppercase.
- The separator must be an underscore.
- The length must be correct.

**Comment by Pokechu22** [comment]
I'm going to have to ask that this be reconsidered.

* Vanilla uses non-standard locales.  Upside-down English (`en_ud`), LOLCAT (`lol_us`), Klingon (`tlh_aa`), etc.  Those need to be handled correctly.
* Resource packs sometimes use non-standard locales too.  This is less frequent, but happens.  You could hope that they fit within the normal rules, but unfortunately that isn't guaranteed.
* I will need to double-check, but I think since 1.11, over the network locales use names like `en_us` and not `en_US`.  The files themselves use those names for sure, as does the setting in `options.txt`.

Simply put, using java's `Locale` class may not work right in all situations, and probably should be avoided.  I haven't actually tested this ingame (I've tried for a bit, but haven't gotten the development environment working quite right) so I'm not 100% sure as to what is/is not broken, but it's still important to consider.

**Comment by kashike** [comment]
We handle it fine currently. https://github.com/SpongePowered/SpongeCommon/blob/bleeding/src/main/java/org/spongepowered/common/util/LocaleCache.java#L40
If we get a real-world report of an issue, I'll gladly fix it. I won't change things because someone wants to use `MadeUp` as a locale. They're going to break more than just Sponge if they do so.
